### ISSUE TITLE
OpenHandler and ResourceProvider for git

### ISSUE THREAD

**Issue by svenefftinge** [theia-ide/theia#505]
As part of #165, we need to register an ResourceProvider

That can be used to open editors and provide resource contents for files from specific git revisions.
For the URI I propose the following:
```
gitrev://repo_name/my/file.tsx#HEAD
\_____/\__________/\__________/\______/
   |         |          |           | 
scheme   repository    path      commitish
```

**Comment by hexa00** [comment]
I think handling #HEAD might be tricky since a file can end with #...

Rather then mitigate this maybe we could have:

```
gitrev://repo_name/commithash/my/file.ts 
```
Then we always use the same separator...

Github does somthing similar like:
```
https://github.com/theia/blob/db5745a1089fdbfb73c12ac90165bd70dfe1b2a7/packages/terminal/src/browser/terminal-widget.ts
```
to view a file

**Comment by svenefftinge** [comment]
Files ending with `#` will cause problems with URIs in general. It is not limited to this style of uri.
But if we require a `gitrev` to come with a fragment, it would even work with hashes in file names, as only the last hash is considered (AFAIR)

**Comment by svenefftinge** [comment]
But thinking about it, as the whole repo is available for a certain revision, we could indeed make it the first segment after the repo.

**Comment by akosyakov** [comment]
In URI query params are always starts from `?`: https://en.wikipedia.org/wiki/Query_string

**Comment by akosyakov** [comment]
In the file path `?` should be encoded. Our URI takes care about it.

**Comment by svenefftinge** [comment]
Fixed with #392
